### DISCUSSION TITLE
How to addresident() method and getonlineplayers() townyuniverse

### DISCUSSION THREAD

**Discussion by JunkieCraft** [TownyAdvanced/Towny#5917] [Q&A]
Hello! I don't know if this is the correct way to post this

I'm trying to remake an old plugin using TownyAPI, I want to add a resident inside a town and get onlinePlayers, but have a couple of errors:

townToBuy = getTownByName(args[1]);
                        try {
                            townToBuy.**addResident(resident)**; **<--Here The method addResident(Resident) from the type Town is not visible**
                            townToBuy.setMayor(resident);
                            townToBuy.getMayor().updatePerms();
                            trade(player, this.database.getData(args[1]).getFileConfiguration().getString("mayorUUID"),
                                    this.database.getData(args[1]).getFileConfiguration().getDouble("price"));
                            this.database.delete(args[1]);
                            TownyUniverse.**getOnlinePlayers(townToBuy)**.forEach(players -> players.sendTitle(

and here: he method **getOnlinePlayers(Town)** is undefined for the type TownyUniverse

variable resident: resident = getResident(player);

I'm making a mistake? Using deprecated methods or incorrects methods? 

Thanks!!

**Comment by LlmDl** [comment]
use resident.setTown(town, true) to add a player to a town.

use TownyAPI.getInstance().getOnlinePlayer(town) for the method that used to be in the TownyUniverse.

**Comment by JunkieCraft** [comment]
Now all variables have an error Unhandled exception type NotRegisteredException:
townToBuy = **getTownByName(args[1]);**
try {
                            resident.setTown(townToBuy);
resident = **getResident(player);**

all variables are inside case "x":  ex::

case "X":
                        resident = **getResident(player)**;
                        if (resident.hasNation()) {
                            player.sendMessage(Main.getChatUtils().getMessage("hasNation"));
                            return true;
                        }
                        if (resident.hasTown()) {
                            player.sendMessage(Main.getChatUtils().getMessage("hasTown"));
                            return true;
                        }
                        if (!this.database.contains(args[1])) {
                            player.sendMessage(Main.getChatUtils().getMessage("townNotFound"));
                            return true;
                        }
                        if (Main.getEconomy().getBalance((OfflinePlayer) player) < this.database.getData(args[1]).getFileConfiguration().getDouble("price")) {
                            player.sendMessage(Main.getChatUtils().getMessage("notEnoughMoney"));
                            return true;
                        }
                        if (player.getName().equals(this.database.getData(args[1]).getFileConfiguration().getString("mayorName"))) {
                            player.sendMessage(Main.getChatUtils().getMessage("youIsOwner"));
                            return true;
                        }
                        townToBuy = **getTownByName(args[1]**);
                        try {
                            resident.setTown(townToBuy);
                            townToBuy.setMayor(resident);
                            townToBuy.getMayor().updatePerms();
                            trade(player, this.database.getData(args[1]).getFileConfiguration().getString("mayorUUID"),
                                    this.database.getData(args[1]).getFileConfiguration().getDouble("price"));
                            this.database.delete(args[1]);
                            TownyAPI.getInstance().getOnlinePlayers(townToBuy).forEach(players -> players.sendTitle(
                                    Main.getChatUtils().getMessage("buyTownLine1").replace("%mayor%", player.getName()),
                                    Main.getChatUtils().getMessage("buyTownLine2").replace("%mayor%", player.getName()),
                                    20, 40, 20));
                        } catch (TownyException e) {
                            e.printStackTrace();
                        }
                        return true;


hmmm >:( idk where the error could be

**Comment by LlmDl** [comment]
Try surrounding your code with 
\```java
code goes here 
\``` 
so that it is readable.

**Comment by JunkieCraft** [comment]
oux sorry!
```java
                 case "X":
                        resident = getResident(player); //Unhandled exception type NotRegisteredException
                        if (resident.hasNation()) {
                            player.sendMessage(Main.getChatUtils().getMessage("hasNation"));
                            return true;
                        }
                        if (resident.hasTown()) {
                            player.sendMessage(Main.getChatUtils().getMessage("hasTown"));
                            return true;
                        }
                        if (!this.database.contains(args[1])) {
                            player.sendMessage(Main.getChatUtils().getMessage("townNotFound"));
                            return true;
                        }
                        if (Main.getEconomy().getBalance((OfflinePlayer) player) < this.database.getData(args[1]).getFileConfiguration().getDouble("price")) {
                            player.sendMessage(Main.getChatUtils().getMessage("notEnoughMoney"));
                            return true;
                        }
                        if (player.getName().equals(this.database.getData(args[1]).getFileConfiguration().getString("mayorName"))) {
                            player.sendMessage(Main.getChatUtils().getMessage("youIsOwner"));
                            return true;
                        }
                        townToBuy = getTownByName(args[1]); //Unhandled exception type NotRegisteredException
                        try {
                            resident.setTown(townToBuy);
                            townToBuy.setMayor(resident);
                            townToBuy.getMayor().updatePerms();
                            trade(player, this.database.getData(args[1]).getFileConfiguration().getString("mayorUUID"),
                                    this.database.getData(args[1]).getFileConfiguration().getDouble("price"));
                            this.database.delete(args[1]);
                            TownyAPI.getInstance().getOnlinePlayers(townToBuy).forEach(players -> players.sendTitle(
                                    Main.getChatUtils().getMessage("buyTownLine1").replace("%mayor%", player.getName()),
                                    Main.getChatUtils().getMessage("buyTownLine2").replace("%mayor%", player.getName()),
                                    20, 40, 20));
                        } catch (TownyException e) {
                            e.printStackTrace();
                        }
                        return true;
```

**Reply by LlmDl** [reply]
How old of a version of Towny is in your plugin's pom.xml?

We don't use the NotRegisteredException very often now, only legacy methods need to use them. Getting a resident from a player should not trigger a NRE if you're on modern Towny versions.

You might want to join the discord, we can figure this out in #plugin-development (just ask for the role when you join the discord,) and then you can post your solution here afterwards for the people that might end up here in the future.

**Comment by JunkieCraft** [comment]
yep is strange Im using the last I think 0.98 in pom.xml.
I will join dc server ans ask for a role

Thanks LlmDl

**Comment by ggvbo** [comment]
Were you able to solve this issue?
